### PR TITLE
feat(exposeBinding): a more powerful exposeFunction w/ attribution

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -421,20 +422,55 @@ will be closed.
...
+
+The first argument of the `playwrightBinding` function contains information about the caller:
+`{ browserContext: BrowserContext, page: Page, frame: Frame }`.
+
+See [page.exposeBinding(name, playwrightBinding)](#pageexposebindingname-playwrightbinding) for page-only version.
+
+An example of exposing page URL to all frames in all pages in the context:
+```js
+const { webkit } = require('playwright');  // Or 'chromium' or 'firefox'.
+const crypto = require('crypto');
```
Comment (yury-s) [inline docs/api.md:442]: Remove this import

File: docs/api.md [hunk 2]
```diff
@@ -1165,13 +1202,52 @@ console.log(await resultHandle.jsonValue());
...
+`{ browserContext: BrowserContext, page: Page, frame: Frame }`.
+
+See [browserContext.exposeBinding(name, playwrightBinding)](#browsercontextexposebindingname-playwrightbinding) for the context-wide version.
+
+> **NOTE** Functions installed via `page.exposeBinding` survive navigations.
+
+An example of exposing page URL to all frames in a page:
+```js
+const { webkit } = require('playwright');  // Or 'chromium' or 'firefox'.
+const crypto = require('crypto');
```
Comment (yury-s) [inline docs/api.md:1224]: Ditto.

File: src/frames.ts [hunk 3]
```diff
@@ -46,6 +47,8 @@ export type GotoResult = {
 
 type ConsoleTagHandler = () => void;
 
+export type FunctionWithSource = (source: { context: BrowserContext, page: Page, frame: Frame}, ...args: any) => any;
```
Comment (yury-s) [inline src/frames.ts:50]: Having a single parameter with named accessors { source, args } could be better and more aligned with other APIs where we allow only single argument. Both approaches are non-ideal though, I'd look for a more elegant way before making this publicly available.

File: docs/api.md [hunk 4]
```diff
@@ -1165,13 +1202,52 @@ console.log(await resultHandle.jsonValue());
 await resultHandle.dispose();
 ```
 
+#### page.exposeBinding(name, playwrightBinding)
```
Comment (yury-s) [inline docs/api.md:1205]: Naming it `exposeBinding` makes it look very different from `exposeFunction` and will likely cause confusion, I'd consider a more similar name `exposeFunctionWithSource` or something, wdyt?

**Review by yury-s — APPROVED** [review]

File: docs/api.md (hunk 4 above)
Comment (pavelfeldman) [reply docs/api.md:1205]: Thought about it, but decided to be more neutral - what if we add more data there later...

File: src/frames.ts (hunk 3 above)
Comment (pavelfeldman) [reply src/frames.ts:50]: Agreed, thought about the same thing

File: docs/api.md (hunk 1 above)
Comment (pavelfeldman) [reply docs/api.md:442]: Done

File: docs/api.md (hunk 2 above)
Comment (pavelfeldman) [reply docs/api.md:1224]: Done

**Pull request merged**